### PR TITLE
Enable sdp_forward for NestedTensors

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This PR implements a sdp_forward for NestedTensors. This impl will call into flash and mem_efficient_attention when possible.

### CODE REVIEW

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 1]
```diff
@@ -10,20 +10,14 @@
...
-
 #include <ATen/native/NonSymbolicBC.h>
 #include <ATen/native/nested/NestedTensorTransformerFunctions.h>
 #include <ATen/native/nested/NestedTensorMath.h>
 #include <ATen/native/nested/NestedTensorUtils.h>
+#include <ATen/native/transformers/cuda/sdp_utils.h>
 
 #include <ATen/cuda/CUDAContext.h>
-
+#include<iostream>
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:20]: nit: did you leave this by accident?

Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:20]: yup

File: aten/src/ATen/native/nested/NestedTensorUtils.cpp [hunk 2]
```diff
@@ -87,14 +87,12 @@ std::vector<Tensor> chunk_nested_tensor(const Tensor& self, int64_t chunks, int6
       // This copys offsets so we are safe to move
       auto new_offsets = std::vector<int64_t>(offsets);
       int64_t *size_ptr = new_sizes.data_ptr<int64_t>();
-      int64_t *stride_ptr = new_strides.data_ptr<int64_t>();
       // Get start val for each split
       int64_t start_val = split_idx * split_size;
       for (int64_t i : c10::irange(n_tensors)) {
         const int64_t index = i * tensor_dim + dim;
-        new_offsets[i] = offsets[i] + start_val * stride_ptr[index];
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorUtils.cpp:95]: What happened here?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 3]
```diff
@@ -218,36 +211,201 @@ Tensor NestedTensor_to_padded_tensor_cuda(
...
+  const int64_t max_seqlen_batch_k = std::get<1>(cumulative_and_max_k);
+
+  const int64_t Nnz_q  = cumulative_sequence_length_q[-1].item<int64_t>();
+  const int64_t Nnz_kv = cumulative_sequence_length_k[-1].item<int64_t>();
+
+  Tensor query_buffer_reshaped;
+  Tensor key_buffer_reshaped;
+  Tensor value_buffer_reshaped;
+
+  if (is_chunked_nested_tensor_group(query, key, value)) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:304]: I don't think this is enough of a condition to ensure that the `as_strided` calls yield correct output.

We don't need to check that query, key and value Storage are all aliases of each other, we just need to check whether that call to `as_strided` is allowed.

Comment (cpuhrsch) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:304]: That means making sure the strides `{3 * num_heads * head_dim, head_dim, 1}` and offset `x * packed_offset` yield a valid view of the data.

File: aten/src/ATen/native/transformers/transformer.cpp [hunk 4]
```diff
@@ -106,7 +106,7 @@ Tensor transformer_encoder_layer_forward(
      x = at::linear(x, qkv_weight, qkv_bias);
      auto x_size_0 = x.size(0);
      x = x.view({x_size_0, -1, 3, num_heads, embed_dim / num_heads});
-     x = flash_attention_helper(x, x, x, 0.0, false, false);
+     x = std::get<0>(flash_attention_helper_packed(x, x, x, 0.0, false, false));
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/transformer.cpp:109]: This is the only call site, why update the signature to return a tuple?

File: aten/src/ATen/native/nested/NestedTensorUtils.cpp (hunk 2 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/NestedTensorUtils.cpp:95]: Since we check that the input is contiguous  stride_ptr[index] which is the stride of the last dim in n_tensor[i] will always equal 1. Was working through this when going over the strides calc for as_strided. I can undo but it is the same resulting calc

File: aten/src/ATen/native/transformers/transformer.cpp (hunk 4 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/transformer.cpp:109]: yeah this was copied in from another PR I will undo

File: test/test_transformers.py [hunk 5]
```diff
@@ -870,6 +870,80 @@ def _test_fastpath(model, mask, mock_return_value, nested_tensors=True):
...
+            actual = torch.nn.functional._scaled_dot_product_attention(
+                query, key, value, attn_mask=None, dropout_p=0.0, need_attn_weights=False, is_causal=False)
+        with sdp_kernel(enable_flash=False):
+            math_ref = torch.nn.functional._scaled_dot_product_attention(
+                query.contiguous(), key.contiguous(), value.contiguous(),
+                attn_mask=None, dropout_p=0.0, need_attn_weights=False, is_causal=False)
+
+        self.assertEqual(actual[0].contiguous(), math_ref[0].contiguous(), atol=1e-3, rtol=1e-2)
+
+    @unittest.skipIf(not SM80OrLater, "Requires SM80 GPU capabailities")
```
Comment (cpuhrsch) [inline test/test_transformers.py:913]: But it shouldn't though

File: test/test_transformers.py [hunk 6]
```diff
@@ -870,6 +870,90 @@ def _test_fastpath(model, mask, mock_return_value, nested_tensors=True):
...
+            actual = torch.nn.functional._scaled_dot_product_attention(
+                query, key, value, attn_mask=None, dropout_p=0.0, need_attn_weights=False, is_causal=False)
+        with sdp_kernel(enable_flash=False):
+            math_ref = torch.nn.functional._scaled_dot_product_attention(
+                query.contiguous(), key.contiguous(), value.contiguous(),
+                attn_mask=None, dropout_p=0.0, need_attn_weights=False, is_causal=False)
+
+        self.assertEqual(actual[0].contiguous(), math_ref[0].contiguous(), atol=1e-3, rtol=1e-2)
+
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
```
Comment (cpuhrsch) [inline test/test_transformers.py:918]: nit: You could use the "gpuonly" decorators for tests. `@onlyCUDA`

Comment (drisspg) [reply test/test_transformers.py:918]: yeah I don't like this either. I tried adding this to the test cases but get segfaults :(  will maybe investigate a little more but kind of copy pasta here. I think though these tests should get cleaned up

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 7]
```diff
@@ -218,36 +213,307 @@ Tensor NestedTensor_to_padded_tensor_cuda(
...
+  cumulative_seqlen = cumulative_seqlen.to(TensorOptions().device(at::kCUDA));
+  return std::tuple<Tensor, int64_t>{cumulative_seqlen, max_seqlen};
+}
+/**
+ * This function checks if the input query, key, value tensors are valid for
+ * use with the flash-attention and efficient_attention kernels without
+ * needing to call contiguous on the nested tensor input.
+ * It checks that the storage offsets' adjacent_difference constant over each of the
+ * tensor in the nested tensor and that the strides are monitonically decreasing.
+ * This is check is done after calling transpose on the nested tensor.
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:284]: nit: "This is check is done"

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 8]
```diff
@@ -218,36 +213,307 @@ Tensor NestedTensor_to_padded_tensor_cuda(
...
+ *
+ * @return A boolean indicating of contiguous needs to be called for input
+ */
+bool is_safe_to_get_storage_as_tensor(
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:288]: Why not do it for one NT at a time? I don't think you need to process all three at once.

File: aten/src/ATen/native/transformers/transformer.cpp [hunk 9]
```diff
@@ -106,7 +106,7 @@ Tensor transformer_encoder_layer_forward(
      x = at::linear(x, qkv_weight, qkv_bias);
      auto x_size_0 = x.size(0);
      x = x.view({x_size_0, -1, 3, num_heads, embed_dim / num_heads});
-     x = flash_attention_helper(x, x, x, 0.0, false, false);
+     x = flash_attention_helper_packed(x, x, x, 0.0, false, false);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/transformer.cpp:109]: Do we need to change the name? Let's just avoid this file entirely.

File: test/test_transformers.py (hunk 6 above)
Comment (cpuhrsch) [reply test/test_transformers.py:918]: We can do that as a follow up

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 10]
```diff
@@ -585,7 +585,7 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_attention_forward_cuda(
     }
 }
 
-std::tuple<Tensor, Tensor> flash_scaled_dot_product_attention(
+Tensor flash_scaled_dot_product_attention(
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention.cu:588]: Do we need to update this function as well? It seems to be only used by `flash_attention_helper_packed`, which is unused in this PR.

**Review by cpuhrsch — APPROVED** [review]

**Comment by cpuhrsch** [comment]
Approved, but please resolve comments before merge.

File: aten/src/ATen/native/transformers/cuda/attention.cu (hunk 10 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/attention.cu:588]: The problem was that there was another implementation of flash_scaled_dot_prod_attention that was in nestedtransformers.cu and this one wasn't getting picked up on. So now there is one and this is it. It matches the one from nestedtransformers.cu

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp (hunk 8 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:288]: So I was wrong about the adjacent differences being constant throughout a tensor because the ragged dimension means the #of elements is going different as you jump through ntensor[i] to ntensor[i+1]. This currently checks that the adjacent differences between q,k,v though matched which would occur if q,k,v were chunked. I wonder if the only condition needed is consistent strides and that they are monotonically decreasing, pseudo contiguous. 

``` c++
q_storage_as_tensor.as_strided(
      {Nnz_q, num_heads, head_dim},
      {nnz_q_stride, head_q_stride, head_dim_stride},
      query_impl->get_storage_offsets()[0]);
```

nnz_q_stride, heqd_q_stride, head_dim_stride would be analogous to the "consistent. last dim" in this case. The ordering implies that the data is in fact laid out in batch, seq_len, num_heads, head_dim in memory.

The adjacent differences between tensors is trying to determine if the offset aspect is valid. We looking for the analogous "consistent offsets". I think the real condition would be that the adjacent difference for a singular nested_tensor is:

nt_offsets[i+1] - nt_offsets[i] = k * numel[i]

if this tensor is the result of chunk than k will be = #chunks
